# Biographic Movie Review Analysis Using Machine Learning for NLP

*Student Id :vcu526*

*Name: Vishalakshi Arumugam*
    

In [1]:
# Import The libraries Needed
import requests
from selenium import webdriver
from selenium.webdriver.support import ui
from selenium.common.exceptions import NoSuchElementException
from selenium.common.exceptions import ElementNotInteractableException
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By
from urllib.parse import urljoin
from bs4 import BeautifulSoup 
import csv
import pandas as pd
import numpy as np 
import re
import string
from imdb import IMDb
import seaborn as sns
# from keras.preprocessing.text import Tokenizer
# from keras.preprocessing.sequence import pad_sequences
# from keras.utils import to_categorical
# import matplotlib.pyplot as plt
# %matplotlib inline
# import spacy

# Scrape the Movie Reviews

For this Analysis I have scraped the Biographic Movie reviews from IMDB Dataset for the Movie whose release year is greater than or equal to 2010.

In [2]:
# Set the Header Details for Movie review File
header1 = ["Movie_Name","Review_Title", "Review_User","Review_content","Rating"]
myfile1 = open("Review.csv", 'w',newline="",encoding='utf-8')
csvwriter = csv.writer(myfile1, delimiter =',')
csvwriter.writerow(header1)
myfile1.close()

In [3]:
# Set up the Urls for the scraping
base_url="https://www.imdb.com"
review_concat="?ref_=undefined&paginationKey="
review_ajax ="/_ajax"
next_url = "https://www.imdb.com/title/tt10618286/reviews/_ajax?ref_=undefined&paginationKey=g4wp7cblr42tazqk7gtxfmrtrtumqbzhzfmxvlnomwklyczuf43o6ss5oi4fxpridb4k54edkxaagcbinoedrsj6xslipfy"


**From IMDB Website have already scraped all the biographic Movies along with the year,Genre, Director, runtime, rating, metascore, review_url. In the Some of the Movie doesnt have release year. Filter out the Movie which doent have proper year and grouped all the movies into a year wise dataframe which in turn used to scrape the movie reviews**

In [17]:
movie_df = pd.read_csv('MovieReview.csv')
movie_df.columns= ['Movie_name','Year','Genre','Director','Runtime','Rating','Metascore','Review_url','Num_of_Votes']
is_numeric_year = movie_df['Year'].astype(str).str.isnumeric()
movie_with_proper_year= movie_df[is_numeric_year]
movie_with_proper_year['Year'].astype(int)
movie_2010_2014_df=movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)>=2010)& (movie_with_proper_year['Year'].astype(int) <=2014)]
movie_2010_2014_df=movie_2010_2014_df.reset_index(drop=True)
print("Index of movie_2010_2014_df:",movie_2010_2014_df.index)
print("movie_2010_2014_df:",len(movie_2010_2014_df))
movie_2015_2017_df=movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)>=2015)& (movie_with_proper_year['Year'].astype(int) <=2017)]
movie_2015_2017_df=movie_2015_2017_df.reset_index(drop=True)
print("Index of movie_2015_2017_df:",movie_2015_2017_df.index)
print("movie_2015_2017_df:",len(movie_2015_2017_df))
movie_2017_till_df=movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)>2017)]
movie_2017_till_df=movie_2017_till_df.reset_index(drop=True)
print("Index of movie_2017_till_df:",movie_2017_till_df.index)
print("movie_2017_till_df:",len(movie_2017_till_df))
movie_2010_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2010)].reset_index(drop=True)
print("Movie_2010_df:",len(movie_2010_df))
print("Index of Movie_2010_df:",movie_2010_df.index)
movie_2011_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2011)].reset_index(drop=True)
print("Movie_2011_df:",len(movie_2011_df))
print("Index of Movie_2011_df:",movie_2011_df.index)
movie_2012_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2012)].reset_index(drop=True)
print("Movie_2012_df:",len(movie_2012_df))
print("Index of Movie_2012_df:",movie_2012_df.index)
movie_2013_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2013)].reset_index(drop=True)
print("Movie_2013_df:",len(movie_2013_df))
print("Index of Movie_2013_df:",movie_2013_df.index)
movie_2014_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2014)].reset_index(drop=True)
print("Movie_2014_df:",len(movie_2014_df))
print("Index of Movie_2014_df:",movie_2014_df.index)
movie_2015_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2015)].reset_index(drop=True)
print("Movie_2015_df:",len(movie_2015_df))
print("Index of Movie_2015_df:",movie_2015_df.index)
movie_2016_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2016)].reset_index(drop=True)
print("Movie_2016_df:",len(movie_2016_df))
print("Index of Movie_2016_df:",movie_2016_df.index)
movie_2017_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2017)].reset_index(drop=True)
print("Movie_2017_df:",len(movie_2017_df))
print("Index of Movie_2017_df:",movie_2017_df.index)
movie_2018_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2018)].reset_index(drop=True)
print("Movie_2018_df:",len(movie_2018_df))
print("Index of Movie_2018_df:",movie_2018_df.index)
movie_2019_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2019)].reset_index(drop=True)
print("Movie_2019_df:",len(movie_2019_df))
print("Index of Movie_2019_df:",movie_2019_df.index)
movie_2020_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2020)].reset_index(drop=True)
print("Movie_2020_df:",len(movie_2020_df))
print("Index of Movie_2020_df:",movie_2020_df.index)
movie_2021_df = movie_with_proper_year[(movie_with_proper_year['Year'].astype(int)==2021)].reset_index(drop=True)
print("Movie_2021_df:",len(movie_2021_df))
print("Index of Movie_2021_df:",movie_2021_df.index)

Index of movie_2010_2014_df: RangeIndex(start=0, stop=868, step=1)
movie_2010_2014_df: 868
Index of movie_2015_2017_df: RangeIndex(start=0, stop=746, step=1)
movie_2015_2017_df: 746
Index of movie_2017_till_df: RangeIndex(start=0, stop=848, step=1)
movie_2017_till_df: 848
Movie_2010_df: 149
Index of Movie_2010_df: RangeIndex(start=0, stop=149, step=1)
Movie_2011_df: 150
Index of Movie_2011_df: RangeIndex(start=0, stop=150, step=1)
Movie_2012_df: 148
Index of Movie_2012_df: RangeIndex(start=0, stop=148, step=1)
Movie_2013_df: 211
Index of Movie_2013_df: RangeIndex(start=0, stop=211, step=1)
Movie_2014_df: 210
Index of Movie_2014_df: RangeIndex(start=0, stop=210, step=1)
Movie_2015_df: 223
Index of Movie_2015_df: RangeIndex(start=0, stop=223, step=1)
Movie_2016_df: 269
Index of Movie_2016_df: RangeIndex(start=0, stop=269, step=1)
Movie_2017_df: 254
Index of Movie_2017_df: RangeIndex(start=0, stop=254, step=1)
Movie_2018_df: 248
Index of Movie_2018_df: RangeIndex(start=0, stop=248, step=1

**Scraped top 25 reviews for the Biographic movies whose release year greater than 2010 and placed in the file named "Review_All_Movie_top25".**

In [10]:
# Code To scrape the 25 reviews for every Biography movie whose release year > 2010
print("Scrape 2010-2014 movies review")
movie_count = 0
for index, movie in movie_2010_2014_df.iterrows():
    print("index:", index)
    print("length:",len(movie_2010_2014_df))
    movie_count+=1
    if index < len(movie_2010_2014_df):
        movie_name_file = movie['Movie_name']
        print('MovieCount:',movie_count)
        print('Movie Name:',movie_name_file)
        Review_page_response,review_content = get_webpage_content(movie['Review_url'])
        Rev_res_code=int(re.findall(r"[0-9]+", str(Review_page_response))[0])
        print(Rev_res_code)
        if (Rev_res_code>=200 | Rev_res_code<300):
            get_review_details(movie_name_file,review_content)
        else:
            print("No Review Available")
print("Scrape 2015-2017 movies review")
movie_count = 0
for index, movie in movie_2015_2017_df.iterrows():
    print("index:", index)
    print("length:",len(movie_2015_2017_df))
    movie_count+=1
    if index < len(movie_2015_2017_df):
        movie_name_file = movie['Movie_name']
        print('MovieCount:',movie_count)
        print('Movie Name:',movie_name_file)
        Review_page_response,review_content = get_webpage_content(movie['Review_url'])
        Rev_res_code=int(re.findall(r"[0-9]+", str(Review_page_response))[0])
        print(Rev_res_code)
        if (Rev_res_code>=200 | Rev_res_code<300):
            get_review_details(movie_name_file,review_content)
        else:
            print("No Review Available")
print("Scrape > 2017 movies review")
movie_count = 0
for index, movie in movie_2017_till_df.iterrows():
    movie_count+=1
    print("index:", index)
    print("length:",len(movie_2017_till_df))
    if index < len(movie_2017_till_df):
        movie_name_file = movie['Movie_name']
        print('MovieCount:',movie_count)
        print('Movie Name:',movie_name_file)
        Review_page_response,review_content = get_webpage_content(movie['Review_url'])
        Rev_res_code=int(re.findall(r"[0-9]+", str(Review_page_response))[0])
        print(Rev_res_code)
        if (Rev_res_code>=200 | Rev_res_code<300):
            get_review_details(movie_name_file,review_content)
        else:
            print("No Review Available")

Scrape 2010-2014 movies review
index: 0
length: 868
MovieCount: 1
Movie Name: The Wolf of Wall Street
200
index: 1
length: 868
MovieCount: 2
Movie Name: Moneyball
200
index: 2
length: 868
MovieCount: 3
Movie Name: The Social Network
200
index: 3
length: 868
MovieCount: 4
Movie Name: Rush
200
index: 4
length: 868
MovieCount: 5
Movie Name: The Intouchables
200
index: 5
length: 868
MovieCount: 6
Movie Name: 12 Years a Slave
200
index: 6
length: 868
MovieCount: 7
Movie Name: Argo
200
index: 7
length: 868
MovieCount: 8
Movie Name: Unbroken
200
index: 8
length: 868
MovieCount: 9
Movie Name: The Imitation Game
200
index: 9
length: 868
MovieCount: 10
Movie Name: Dallas Buyers Club
200
index: 10
length: 868
MovieCount: 11
Movie Name: 42
200
index: 11
length: 868
MovieCount: 12
Movie Name: Lincoln
200
index: 12
length: 868
MovieCount: 13
Movie Name: American Sniper
200
index: 13
length: 868
MovieCount: 14
Movie Name: Lone Survivor
200
index: 14
length: 868
MovieCount: 15
Movie Name: The King's S

200
index: 121
length: 868
MovieCount: 122
Movie Name: The Hammer
200
index: 122
length: 868
MovieCount: 123
Movie Name: Jimi: All Is by My Side
200
index: 123
length: 868
MovieCount: 124
Movie Name: The Dark Horse
200
index: 124
length: 868
MovieCount: 125
Movie Name: Pasolini
200
index: 125
length: 868
MovieCount: 126
Movie Name: Gloria
200
index: 126
length: 868
MovieCount: 127
Movie Name: The Devil's Violinist
200
index: 127
length: 868
MovieCount: 128
Movie Name: The Legend Is Born: Ip Man
200
index: 128
length: 868
MovieCount: 129
Movie Name: Rosewater
200
index: 129
length: 868
MovieCount: 130
Movie Name: Eden
200
index: 130
length: 868
MovieCount: 131
Movie Name: Jimmy's Hall
200
index: 131
length: 868
MovieCount: 132
Movie Name: Against the Sun
200
index: 132
length: 868
MovieCount: 133
Movie Name: Digging Up the Marrow
200
index: 133
length: 868
MovieCount: 134
Movie Name: Mary Queen of Scots
200
index: 134
length: 868
MovieCount: 135
Movie Name: Reaching for the Moon
200
ind

200
index: 238
length: 868
MovieCount: 239
Movie Name: Mastram
200
index: 239
length: 868
MovieCount: 240
Movie Name: El Presidente
200
index: 240
length: 868
MovieCount: 241
Movie Name: Difret
200
index: 241
length: 868
MovieCount: 242
Movie Name: Mozart's Sister
200
index: 242
length: 868
MovieCount: 243
Movie Name: Marie's Story
200
index: 243
length: 868
MovieCount: 244
Movie Name: Louis Cyr
200
index: 244
length: 868
MovieCount: 245
Movie Name: Underground: The Julian Assange Story
200
index: 245
length: 868
MovieCount: 246
Movie Name: 11.6
200
index: 246
length: 868
MovieCount: 247
Movie Name: My Paparotti
200
index: 247
length: 868
MovieCount: 248
Movie Name: Rengô kantai shirei chôkan: Yamamoto Isoroku
200
index: 248
length: 868
MovieCount: 249
Movie Name: Colette
200
index: 249
length: 868
MovieCount: 250
Movie Name: Présumé coupable
200
index: 250
length: 868
MovieCount: 251
Movie Name: Sri Siddhartha Gautama
200
index: 251
length: 868
MovieCount: 252
Movie Name: Jaatishwar
2

200
index: 352
length: 868
MovieCount: 353
Movie Name: Beduin
200
index: 353
length: 868
MovieCount: 354
Movie Name: Getúlio
200
index: 354
length: 868
MovieCount: 355
Movie Name: King of the Sands
200
index: 355
length: 868
MovieCount: 356
Movie Name: Luka
200
index: 356
length: 868
MovieCount: 357
Movie Name: Falks Grav
200
index: 357
length: 868
MovieCount: 358
Movie Name: Nainsukh
200
index: 358
length: 868
MovieCount: 359
Movie Name: Chunmeng
200
index: 359
length: 868
MovieCount: 360
Movie Name: James McNeill Whistler and the Case for Beauty
200
index: 360
length: 868
MovieCount: 361
Movie Name: Meghe Dhaka Tara
200
index: 361
length: 868
MovieCount: 362
Movie Name: Josh
200
index: 362
length: 868
MovieCount: 363
Movie Name: Undaunted... The Early Life of Josh McDowell
200
index: 363
length: 868
MovieCount: 364
Movie Name: Benjamin Britten: Peace and Conflict
200
index: 364
length: 868
MovieCount: 365
Movie Name: 476 A.D. Chapter One: The Last Light of Aries
200
index: 365
length

200
index: 463
length: 868
MovieCount: 464
Movie Name: Trinta
200
index: 464
length: 868
MovieCount: 465
Movie Name: The Gertrude Stein Mystery or Some Like It Art
200
index: 465
length: 868
MovieCount: 466
Movie Name: The David Whiting Story
200
index: 466
length: 868
MovieCount: 467
Movie Name: Dominic: Light of the Church
200
index: 467
length: 868
MovieCount: 468
Movie Name: Odbrojavanje
200
index: 468
length: 868
MovieCount: 469
Movie Name: Bolívar el hombre de las dificultades
200
index: 469
length: 868
MovieCount: 470
Movie Name: Zyvie Belarus
200
index: 470
length: 868
MovieCount: 471
Movie Name: Farsh wa Ghata
200
index: 471
length: 868
MovieCount: 472
Movie Name: O Fantasma do Novais
200
index: 472
length: 868
MovieCount: 473
Movie Name: Amy Winehouse: Fallen Star
200
index: 473
length: 868
MovieCount: 474
Movie Name: Steve Jobs: iGenius
200
index: 474
length: 868
MovieCount: 475
Movie Name: Karol który zostal swietym
200
index: 475
length: 868
MovieCount: 476
Movie Name: Zer

200
index: 571
length: 868
MovieCount: 572
Movie Name: Miguel San Miguel
200
index: 572
length: 868
MovieCount: 573
Movie Name: JJ Star... How Embarrassing
200
index: 573
length: 868
MovieCount: 574
Movie Name: Metal Heads
200
index: 574
length: 868
MovieCount: 575
Movie Name: Swabhoomi
200
index: 575
length: 868
MovieCount: 576
Movie Name: The Killers In Connecticut
200
index: 576
length: 868
MovieCount: 577
Movie Name: Ven. Jian Zhen
200
index: 577
length: 868
MovieCount: 578
Movie Name: Perurena
200
index: 578
length: 868
MovieCount: 579
Movie Name: Francis of Brooklyn
200
index: 579
length: 868
MovieCount: 580
Movie Name: Dios Me Libre
200
index: 580
length: 868
MovieCount: 581
Movie Name: El Rayo
200
index: 581
length: 868
MovieCount: 582
Movie Name: Hello Stranger
200
index: 582
length: 868
MovieCount: 583
Movie Name: Ek Noor
200
index: 583
length: 868
MovieCount: 584
Movie Name: Schwarzer Zucker
200
index: 584
length: 868
MovieCount: 585
Movie Name: Persistence of Memory
200
ind

200
index: 678
length: 868
MovieCount: 679
Movie Name: Unauthorized Biography Beyonce: Baby and Beyond
200
index: 679
length: 868
MovieCount: 680
Movie Name: Papa Boss
200
index: 680
length: 868
MovieCount: 681
Movie Name: The Milkweed Tree of Life
200
index: 681
length: 868
MovieCount: 682
Movie Name: The Mystery of Mazo de la Roche
200
index: 682
length: 868
MovieCount: 683
Movie Name: Shqipja E Kavkazit: Imam Shamili
200
index: 683
length: 868
MovieCount: 684
Movie Name: Adrià Puntí
200
index: 684
length: 868
MovieCount: 685
Movie Name: Body and Brain
200
index: 685
length: 868
MovieCount: 686
Movie Name: King of Hustlin
200
index: 686
length: 868
MovieCount: 687
Movie Name: Mikis Theodorakis - O synthetis ton poiiton kai ton oramaton 1950 eos simera
200
index: 687
length: 868
MovieCount: 688
Movie Name: Angel in June
200
index: 688
length: 868
MovieCount: 689
Movie Name: Käshbasshy zholy
200
index: 689
length: 868
MovieCount: 690
Movie Name: Mao Zedong and Qi Baishi
200
index: 690


200
index: 781
length: 868
MovieCount: 782
Movie Name: Lucia Venise
200
index: 782
length: 868
MovieCount: 783
Movie Name: Il vento fra le mura
200
index: 783
length: 868
MovieCount: 784
Movie Name: Arrivederci Macau
200
index: 784
length: 868
MovieCount: 785
Movie Name: Wise Women of Sedona
200
index: 785
length: 868
MovieCount: 786
Movie Name: Anna Teresa e le resistenti
200
index: 786
length: 868
MovieCount: 787
Movie Name: Uno scrittore al fronte
200
index: 787
length: 868
MovieCount: 788
Movie Name: The Tower of Immortality
200
index: 788
length: 868
MovieCount: 789
Movie Name: I Was Saddam's Son
200
index: 789
length: 868
MovieCount: 790
Movie Name: Lady Gaga: On the Edge
200
index: 790
length: 868
MovieCount: 791
Movie Name: La Statue de la Liberté naissance d'un symbole
200
index: 791
length: 868
MovieCount: 792
Movie Name: Ya dumal: ya - Adam... ili Dolgaya doroga v Gogezhu
200
index: 792
length: 868
MovieCount: 793
Movie Name: The Private Life of Marilyn Monroe
200
index: 793

200
index: 18
length: 746
MovieCount: 19
Movie Name: Lion
200
index: 19
length: 746
MovieCount: 20
Movie Name: In the Heart of the Sea
200
index: 20
length: 746
MovieCount: 21
Movie Name: The Lost City of Z
200
index: 21
length: 746
MovieCount: 22
Movie Name: Dangal
200
index: 22
length: 746
MovieCount: 23
Movie Name: All the Money in the World
200
index: 23
length: 746
MovieCount: 24
Movie Name: Jungle
200
index: 24
length: 746
MovieCount: 25
Movie Name: Colonia
200
index: 25
length: 746
MovieCount: 26
Movie Name: Black Mass
200
index: 26
length: 746
MovieCount: 27
Movie Name: Only the Brave
200
index: 27
length: 746
MovieCount: 28
Movie Name: Queen of Katwe
200
index: 28
length: 746
MovieCount: 29
Movie Name: Snowden
200
index: 29
length: 746
MovieCount: 30
Movie Name: The Infiltrator
200
index: 30
length: 746
MovieCount: 31
Movie Name: The Zookeeper's Wife
200
index: 31
length: 746
MovieCount: 32
Movie Name: Sully
200
index: 32
length: 746
MovieCount: 33
Movie Name: The Current War


200
index: 136
length: 746
MovieCount: 137
Movie Name: Miles Ahead
200
index: 137
length: 746
MovieCount: 138
Movie Name: Freeheld
200
index: 138
length: 746
MovieCount: 139
Movie Name: Viceroy's House
200
index: 139
length: 746
MovieCount: 140
Movie Name: Woodlawn
200
index: 140
length: 746
MovieCount: 141
Movie Name: Godard Mon Amour
200
index: 141
length: 746
MovieCount: 142
Movie Name: Down by Love
200
index: 142
length: 746
MovieCount: 143
Movie Name: Screwed
200
index: 143
length: 746
MovieCount: 144
Movie Name: The Admiral
200
index: 144
length: 746
MovieCount: 145
Movie Name: Experimenter
200
index: 145
length: 746
MovieCount: 146
Movie Name: Fathom Premieres 6 Below: Miracle on the Mountain
200
index: 146
length: 746
MovieCount: 147
Movie Name: Pele: Birth of a Legend
200
index: 147
length: 746
MovieCount: 148
Movie Name: Roxanne Roxanne
200
index: 148
length: 746
MovieCount: 149
Movie Name: Neerja
200
index: 149
length: 746
MovieCount: 150
Movie Name: Eksi Elmalar
200
index: 

200
index: 250
length: 746
MovieCount: 251
Movie Name: The Meanest Man in Texas
200
index: 251
length: 746
MovieCount: 252
Movie Name: Addiction: A 60's Love Story
200
index: 252
length: 746
MovieCount: 253
Movie Name: Kazantzakis
200
index: 253
length: 746
MovieCount: 254
Movie Name: Najlepszy
200
index: 254
length: 746
MovieCount: 255
Movie Name: A Real Vermeer
200
index: 255
length: 746
MovieCount: 256
Movie Name: Poorna
200
index: 256
length: 746
MovieCount: 257
Movie Name: Kid Kulafu
200
index: 257
length: 746
MovieCount: 258
Movie Name: Lou Andreas-Salomé The Audacity to be Free
200
index: 258
length: 746
MovieCount: 259
Movie Name: Vico C: La vida del filosofo
200
index: 259
length: 746
MovieCount: 260
Movie Name: Memoria
200
index: 260
length: 746
MovieCount: 261
Movie Name: Gabriel e a Montanha
200
index: 261
length: 746
MovieCount: 262
Movie Name: Memoir of War
200
index: 262
length: 746
MovieCount: 263
Movie Name: Egon Schiele: Tod und Mädchen
200
index: 263
length: 746
Movi

200
index: 364
length: 746
MovieCount: 365
Movie Name: Francisco - El Padre Jorge
200
index: 365
length: 746
MovieCount: 366
Movie Name: El Inca
200
index: 366
length: 746
MovieCount: 367
Movie Name: Pilecki
200
index: 367
length: 746
MovieCount: 368
Movie Name: The Waiting Room
200
index: 368
length: 746
MovieCount: 369
Movie Name: Ekk Albela
200
index: 369
length: 746
MovieCount: 370
Movie Name: ¡Asu Mare! 2
200
index: 370
length: 746
MovieCount: 371
Movie Name: Pequeno Segredo
200
index: 371
length: 746
MovieCount: 372
Movie Name: Luz de Soledad
200
index: 372
length: 746
MovieCount: 373
Movie Name: Zeus
200
index: 373
length: 746
MovieCount: 374
Movie Name: Punk Strut: The Movie
200
index: 374
length: 746
MovieCount: 375
Movie Name: Vincent Van Gogh: A New Way of Seeing
200
index: 375
length: 746
MovieCount: 376
Movie Name: Moya babusya Fani Kaplan
200
index: 376
length: 746
MovieCount: 377
Movie Name: Istadeh Dar Ghobar
200
index: 377
length: 746
MovieCount: 378
Movie Name: Rudolp

200
index: 476
length: 746
MovieCount: 477
Movie Name: Judas
200
index: 477
length: 746
MovieCount: 478
Movie Name: Mondo Americana
200
index: 478
length: 746
MovieCount: 479
Movie Name: La luz con el tiempo dentro
200
index: 479
length: 746
MovieCount: 480
Movie Name: Chaïbia
200
index: 480
length: 746
MovieCount: 481
Movie Name: The Great Leader Kanshiram
200
index: 481
length: 746
MovieCount: 482
Movie Name: Broken Chains
200
index: 482
length: 746
MovieCount: 483
Movie Name: Anna Nicole & the Starlight Angel
200
index: 483
length: 746
MovieCount: 484
Movie Name: Tainted Love
200
index: 484
length: 746
MovieCount: 485
Movie Name: La leggenda di Bob Wind
200
index: 485
length: 746
MovieCount: 486
Movie Name: Alive
200
index: 486
length: 746
MovieCount: 487
Movie Name: Satoshi no seishun
200
index: 487
length: 746
MovieCount: 488
Movie Name: Voyage avec un âne dans les Cévennes
200
index: 488
length: 746
MovieCount: 489
Movie Name: Patriot of the Web
200
index: 489
length: 746
MovieCo

200
index: 584
length: 746
MovieCount: 585
Movie Name: Nel mondo grande e terribile
200
index: 585
length: 746
MovieCount: 586
Movie Name: Esterházy Péter emlékfilm ... a mindiget javítom örökkére
200
index: 586
length: 746
MovieCount: 587
Movie Name: Churchill: The One-Man Show
200
index: 587
length: 746
MovieCount: 588
Movie Name: Priyamanasam
200
index: 588
length: 746
MovieCount: 589
Movie Name: Aspira
200
index: 589
length: 746
MovieCount: 590
Movie Name: Forgive Me Father for I Have Sinned
200
index: 590
length: 746
MovieCount: 591
Movie Name: Robin Schulz: The Movie
200
index: 591
length: 746
MovieCount: 592
Movie Name: Aktoty
200
index: 592
length: 746
MovieCount: 593
Movie Name: Carola Neher - Todesursache unbekannt
200
index: 593
length: 746
MovieCount: 594
Movie Name: Interview wit German
200
index: 594
length: 746
MovieCount: 595
Movie Name: Clint
200
index: 595
length: 746
MovieCount: 596
Movie Name: Un nuovo giorno
200
index: 596
length: 746
MovieCount: 597
Movie Name: Yu

200
index: 688
length: 746
MovieCount: 689
Movie Name: British
200
index: 689
length: 746
MovieCount: 690
Movie Name: Portræt af Peter Belli
200
index: 690
length: 746
MovieCount: 691
Movie Name: kero u a c s on the road
200
index: 691
length: 746
MovieCount: 692
Movie Name: Scandinav But Oscar
200
index: 692
length: 746
MovieCount: 693
Movie Name: Mr.Veerappan
200
index: 693
length: 746
MovieCount: 694
Movie Name: DiCaprio: Long Road to the Oscars
200
index: 694
length: 746
MovieCount: 695
Movie Name: Rosso dal vetro: dovunque tu vada
200
index: 695
length: 746
MovieCount: 696
Movie Name: Ashik
200
index: 696
length: 746
MovieCount: 697
Movie Name: Nei Ngaidam in
200
index: 697
length: 746
MovieCount: 698
Movie Name: harry
200
index: 698
length: 746
MovieCount: 699
Movie Name: League of Heels: A Video Game Wrestling Documentary
200
index: 699
length: 746
MovieCount: 700
Movie Name: Marie's Attitude
200
index: 700
length: 746
MovieCount: 701
Movie Name: The Lion of Penteli
200
index: 7

200
index: 57
length: 848
MovieCount: 58
Movie Name: The Boy Who Harnessed the Wind
200
index: 58
length: 848
MovieCount: 59
Movie Name: Curiosa
200
index: 59
length: 848
MovieCount: 60
Movie Name: Togo
200
index: 60
length: 848
MovieCount: 61
Movie Name: American Animals
200
index: 61
length: 848
MovieCount: 62
Movie Name: Ip Man 4: The Finale
200
index: 62
length: 848
MovieCount: 63
Movie Name: The Best of Enemies
200
index: 63
length: 848
MovieCount: 64
Movie Name: Blonde
200
index: 64
length: 848
MovieCount: 65
Movie Name: The Report
200
index: 65
length: 848
MovieCount: 66
Movie Name: Boy Erased
200
index: 66
length: 848
MovieCount: 67
Movie Name: Son of the South
200
index: 67
length: 848
MovieCount: 68
Movie Name: Tesla
200
index: 68
length: 848
MovieCount: 69
Movie Name: Dolemite Is My Name
200
index: 69
length: 848
MovieCount: 70
Movie Name: The Catcher Was a Spy
200
index: 70
length: 848
MovieCount: 71
Movie Name: Mary Magdalene
200
index: 71
length: 848
MovieCount: 72
Movie 

200
index: 177
length: 848
MovieCount: 178
Movie Name: Bizim Için Sampiyon
200
index: 178
length: 848
MovieCount: 179
Movie Name: Tanhaji: The Unsung Warrior
200
index: 179
length: 848
MovieCount: 180
Movie Name: Walk. Ride. Rodeo.
200
index: 180
length: 848
MovieCount: 181
Movie Name: De Gaulle
200
index: 181
length: 848
MovieCount: 182
Movie Name: Enfant Terrible
200
index: 182
length: 848
MovieCount: 183
Movie Name: Un amour impossible
200
index: 183
length: 848
MovieCount: 184
Movie Name: The Judgement
200
index: 184
length: 848
MovieCount: 185
Movie Name: In the Shadow of the Mountain
200
index: 185
length: 848
MovieCount: 186
Movie Name: Saint Judy
200
index: 186
length: 848
MovieCount: 187
Movie Name: Corazón
200
index: 187
length: 848
MovieCount: 188
Movie Name: Io sono Mia
200
index: 188
length: 848
MovieCount: 189
Movie Name: Hidden Away
200
index: 189
length: 848
MovieCount: 190
Movie Name: Müslüm
200
index: 190
length: 848
MovieCount: 191
Movie Name: Irena Sendler
200
index

200
index: 292
length: 848
MovieCount: 293
Movie Name: General
200
index: 293
length: 848
MovieCount: 294
Movie Name: Zhan shen ji
200
index: 294
length: 848
MovieCount: 295
Movie Name: #FriendButMarried 2
200
index: 295
length: 848
MovieCount: 296
Movie Name: Deadman Standing
200
index: 296
length: 848
MovieCount: 297
Movie Name: Main Mulayam Singh Yadav
200
index: 297
length: 848
MovieCount: 298
Movie Name: Jhund
200
index: 298
length: 848
MovieCount: 299
Movie Name: The Maestro
200
index: 299
length: 848
MovieCount: 300
Movie Name: Tortured for Christ
200
index: 300
length: 848
MovieCount: 301
Movie Name: Goyo: Ang Batang Heneral
200
index: 301
length: 848
MovieCount: 302
Movie Name: Lata Bhagwan Kare
200
index: 302
length: 848
MovieCount: 303
Movie Name: Borderline
200
index: 303
length: 848
MovieCount: 304
Movie Name: The Royal
200
index: 304
length: 848
MovieCount: 305
Movie Name: Zátopek
200
index: 305
length: 848
MovieCount: 306
Movie Name: The Walk-On's
200
index: 306
length: 

200
index: 407
length: 848
MovieCount: 408
Movie Name: Adam Sandler: Funny Guy
200
index: 408
length: 848
MovieCount: 409
Movie Name: Jack Jonah
200
index: 409
length: 848
MovieCount: 410
Movie Name: El Potro lo mejor del amor
200
index: 410
length: 848
MovieCount: 411
Movie Name: Selfie
200
index: 411
length: 848
MovieCount: 412
Movie Name: The Whalebone Box
200
index: 412
length: 848
MovieCount: 413
Movie Name: La estrella roja
200
index: 413
length: 848
MovieCount: 414
Movie Name: Bangabandhu
200
index: 414
length: 848
MovieCount: 415
Movie Name: Hebe: A Estrela do Brasil
200
index: 415
length: 848
MovieCount: 416
Movie Name: Mihkel
200
index: 416
length: 848
MovieCount: 417
Movie Name: Project Kahuta
200
index: 417
length: 848
MovieCount: 418
Movie Name: Divaldo: O Mensageiro da Paz
200
index: 418
length: 848
MovieCount: 419
Movie Name: E. M. Forster: His Longest Journey
200
index: 419
length: 848
MovieCount: 420
Movie Name: A Rush of Hope: Find Answers to Life's Questions
200
inde

200
index: 518
length: 848
MovieCount: 519
Movie Name: Watashi wa zettai yurusanai
200
index: 519
length: 848
MovieCount: 520
Movie Name: Dealings with the Downright Devious
200
index: 520
length: 848
MovieCount: 521
Movie Name: Home Cookin: 5.17.18
200
index: 521
length: 848
MovieCount: 522
Movie Name: Trip
200
index: 522
length: 848
MovieCount: 523
Movie Name: The Hill
200
index: 523
length: 848
MovieCount: 524
Movie Name: O Labirinto da Saudade
200
index: 524
length: 848
MovieCount: 525
Movie Name: Hiralal
200
index: 525
length: 848
MovieCount: 526
Movie Name: Untold the Aaliyah Story Back & Forth
200
index: 526
length: 848
MovieCount: 527
Movie Name: Dr Tatya Lahane Angaar Power is within
200
index: 527
length: 848
MovieCount: 528
Movie Name: Begirada
200
index: 528
length: 848
MovieCount: 529
Movie Name: El universo Montesinos
200
index: 529
length: 848
MovieCount: 530
Movie Name: The Kellie Maloney Story
200
index: 530
length: 848
MovieCount: 531
Movie Name: Shi fang
200
index: 5

200
index: 625
length: 848
MovieCount: 626
Movie Name: Jumprava: Lielais notikums
200
index: 626
length: 848
MovieCount: 627
Movie Name: L'ultimo uomo che dipinse il cinema
200
index: 627
length: 848
MovieCount: 628
Movie Name: Ilk Adim: 1919
200
index: 628
length: 848
MovieCount: 629
Movie Name: Stomping in Puddles
200
index: 629
length: 848
MovieCount: 630
Movie Name: Jacob A. Riis
200
index: 630
length: 848
MovieCount: 631
Movie Name: Naming Names
200
index: 631
length: 848
MovieCount: 632
Movie Name: Babu Bhaijan
200
index: 632
length: 848
MovieCount: 633
Movie Name: The Calling
200
index: 633
length: 848
MovieCount: 634
Movie Name: Birendra Krishna Bhadra
200
index: 634
length: 848
MovieCount: 635
Movie Name: S.P.E.E.D.
200
index: 635
length: 848
MovieCount: 636
Movie Name: Chain of Life: The Rick Boyle Story
200
index: 636
length: 848
MovieCount: 637
Movie Name: The Skin of My Father
200
index: 637
length: 848
MovieCount: 638
Movie Name: Shaheed- E- Aazam
200
index: 638
length: 8

200
index: 735
length: 848
MovieCount: 736
Movie Name: Your Consideration
200
index: 736
length: 848
MovieCount: 737
Movie Name: Vidas Embarcadas
200
index: 737
length: 848
MovieCount: 738
Movie Name: From the Cold Streets to the Hot Lights
200
index: 738
length: 848
MovieCount: 739
Movie Name: In Greene
200
index: 739
length: 848
MovieCount: 740
Movie Name: Soñando Un Lugar
200
index: 740
length: 848
MovieCount: 741
Movie Name: San Telmo Tapes
200
index: 741
length: 848
MovieCount: 742
Movie Name: Guru ji: Ahead of time
200
index: 742
length: 848
MovieCount: 743
Movie Name: Tihi veceri
200
index: 743
length: 848
MovieCount: 744
Movie Name: Do You Love Me (Documentary)
200
index: 744
length: 848
MovieCount: 745
Movie Name: Between Time and Space
200
index: 745
length: 848
MovieCount: 746
Movie Name: Spider Murphy Gang - Glory Days of Rock 'n' Roll
200
index: 746
length: 848
MovieCount: 747
Movie Name: La corda spezzata
200
index: 747
length: 848
MovieCount: 748
Movie Name: The Running 

200
index: 841
length: 848
MovieCount: 842
Movie Name: Windcatcher Dream
200
index: 842
length: 848
MovieCount: 843
Movie Name: Brendan Grace: Funny Man
200
index: 843
length: 848
MovieCount: 844
Movie Name: The Remote Daughter
200
index: 844
length: 848
MovieCount: 845
Movie Name: Anime notturne
200
index: 845
length: 848
MovieCount: 846
Movie Name: Iz nesigurnosti u neizvjesnost
200
index: 846
length: 848
MovieCount: 847
Movie Name: Saturday's Children
200
index: 847
length: 848
MovieCount: 848
Movie Name: Young Fame
200


**Scraping All the Reviews takes longer time . So I have placed the code below to scraped around 120 reviews and placed the file separately for different year grouping. Below are the File Names with Different year Movie reviews.**
1)Review_2010.csv
2)Review_2011.csv
3)Review_2012.csv
4)Review_2013.csv
4)Review_2014.csv
6)Review_2015_2017.csv
7)Review_2018.csv
8)Review_2019.csv
9)Review_2020.csv
10)Review_2021.csv

In [8]:
# Code To scrape the 500 reviews for every Biography movie whose release year > 2010
movie_count=0
for index, movie in movie_2021_df.iterrows():
    movie_count+=1
    if index < len(movie_2021_df):
        print("index:", index)
        print("length:",len(movie_2021_df))
        count = 0
        load_more_url_ajax = set()
        movie_name_file = movie['Movie_name']
        print('MovieCount:',movie_count)
        print('Movie Name:',movie_name_file)
        load_more_url_ajax.add(movie['Review_url']+review_ajax)
        driver = webdriver.Chrome()
        driver.maximize_window()
        driver.get(movie['Review_url'])
        while True:
            try:
                ActionChains(driver).move_to_element(WebDriverWait(driver, 5).until(EC.element_to_be_clickable((By.XPATH, "//*[(@id='load-more-trigger')]")))).pause(5).click().perform()
                print("LOAD MORE RESULTS button clicked")
                soup = BeautifulSoup(driver.page_source,'html.parser')
                count +=1
                load_more_url_ajax.add(check_Load_more(soup))
                if count >5:
                    break
            except ElementNotInteractableException:
                print("No element interactable Reached bottom of page")
                break
            except NoSuchElementException:
                print("no such element Reached bottom of page")
                break
            except TimeoutException:
                print("No more LOAD MORE RESULTS button to be clicked")
                break
        driver.quit()
#         print(load_more_url_ajax)

        for url in load_more_url_ajax:
#             print("url: ", url)
            Review_page_response,review_content = get_webpage_content(url)
            Rev_res_code=int(re.findall(r"[0-9]+", str(Review_page_response))[0])
#             print(Rev_res_code)
            if (Rev_res_code>=200 | Rev_res_code<300):
                get_review_details(movie_name_file,review_content)
            else:
                print("No Review Available")
#     break

index: 0
length: 156
MovieCount: 1
Movie Name: Judas and the Black Messiah
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
index: 1
length: 156
MovieCount: 2
Movie Name: The Big Bull
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
index: 2
length: 156
MovieCount: 3
Movie Name: The Dig
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
LOAD MORE RESULTS button clicked
index: 3
length: 156
MovieCount: 4
Movie Name: Madame Claude
LOAD MORE RESULTS button clicked
No more LOAD MORE RESULTS button to be clicked
index: 4
length: 156
MovieCount: 5
Movie Name: The United States vs. Billie

No more LOAD MORE RESULTS button to be clicked
index: 66
length: 156
MovieCount: 67
Movie Name: La estrella roja
No more LOAD MORE RESULTS button to be clicked
index: 67
length: 156
MovieCount: 68
Movie Name: Bangabandhu
No more LOAD MORE RESULTS button to be clicked
index: 68
length: 156
MovieCount: 69
Movie Name: Project Kahuta
No more LOAD MORE RESULTS button to be clicked
index: 69
length: 156
MovieCount: 70
Movie Name: I See You and You See Me
No more LOAD MORE RESULTS button to be clicked
index: 70
length: 156
MovieCount: 71
Movie Name: Cam
No more LOAD MORE RESULTS button to be clicked
index: 71
length: 156
MovieCount: 72
Movie Name: Tath Kana
No more LOAD MORE RESULTS button to be clicked
index: 72
length: 156
MovieCount: 73
Movie Name: The Earned Life
No more LOAD MORE RESULTS button to be clicked
index: 73
length: 156
MovieCount: 74
Movie Name: Laleh (Drive)
No more LOAD MORE RESULTS button to be clicked
index: 74
length: 156
MovieCount: 75
Movie Name: The Immortal President


No more LOAD MORE RESULTS button to be clicked
index: 137
length: 156
MovieCount: 138
Movie Name: Fauja
No more LOAD MORE RESULTS button to be clicked
index: 138
length: 156
MovieCount: 139
Movie Name: La primera toga
No more LOAD MORE RESULTS button to be clicked
index: 139
length: 156
MovieCount: 140
Movie Name: Foil
No more LOAD MORE RESULTS button to be clicked
index: 140
length: 156
MovieCount: 141
Movie Name: Nil Cokhe
No more LOAD MORE RESULTS button to be clicked
index: 141
length: 156
MovieCount: 142
Movie Name: Mere Jinvar Mere Guruvar
No more LOAD MORE RESULTS button to be clicked
index: 142
length: 156
MovieCount: 143
Movie Name: Diomar
No more LOAD MORE RESULTS button to be clicked
index: 143
length: 156
MovieCount: 144
Movie Name: Qajymuqan
No more LOAD MORE RESULTS button to be clicked
index: 144
length: 156
MovieCount: 145
Movie Name: Your Consideration
No more LOAD MORE RESULTS button to be clicked
index: 145
length: 156
MovieCount: 146
Movie Name: Do You Love Me (Docu

In [7]:
# Get the page content in a HTML Format
def get_webpage_content(url):
    response = requests.get(url)
    page_content = BeautifulSoup(response.text,'html.parser')
    return (response, page_content)

In [8]:
# Get the Ajax Pagination Next Page URL
def check_Load_more(web_content):
    next_page_url = ""
    if web_content.find('div',{'class':'load-more-data'}):
        data_ajaxurl = web_content.find('div',{'class':'load-more-data'})['data-ajaxurl']
        data_key = " "
        try:
            data_key = web_content.find('div',{'class':'load-more-data'})['data-key']
            next_page_url = base_url+data_ajaxurl+review_concat+data_key
        except KeyError:
            next_page_url = " "
            print("end of the page")
        return next_page_url
    else:
        next_page_url =""
        return next_page_url

In [9]:
# Get the Review details 
def get_review_details(movie_name_input,Rev_content):
    Mov_name = movie_name_input
    Review_Det = Rev_content.findAll('div',{'lister-item-content'})
    for i in Review_Det:
        Rev_title= i.find('a').text.strip()
        Rev_title = Rev_title.translate(str.maketrans('','',string.punctuation))
        Rev_user = i.find('span',{'display-name-link','a'}).text
        if i. find('div',class_='text show-more__control'):
            Review_content = i. find('div',class_='text show-more__control').text
            Review_content = Review_content.translate(str.maketrans('','',string.punctuation)) 
            Review_content = Review_content.replace("\n","")
        else:
            Review_content = "Not Applicable"
        if i.find('div',{'ipl-ratings-bar'}):
            Rating = i.find('span').text.strip().split('/')[0]
        else:
            Rating = "NA"
        
        myfile1 =open("Review.csv", 'a', newline="",encoding='utf8', errors = "ignore")
        csvwriter = csv.writer(myfile1)
        myfile1.write(Mov_name+','+Rev_title+','+Rev_user+','+Review_content+','+Rating+','+'\n')
        myfile1.close()

# Import Files with Top 120 Reviews

In [18]:
# Import all the csv file with top 120 reviews and convert that into a dataframe.
# All_movie_review Data set contains 55741 Record
import glob
import os
path ='C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset/'
filenames = glob.glob(os.path.join(path, "*.csv"))
print(filenames)
total = 0
All_movie_review=pd.DataFrame()
for file in filenames:
    print(file)
    
    df = pd.read_csv(file,encoding='utf-8',header=0,usecols=['Movie_Name','Review_Title','Review_User','Review_content','Rating'])
    print(len(df))
    total = total+len(df)
    All_movie_review=All_movie_review.append(df,ignore_index=True)
    
print(len(All_movie_review)) 
print('Total:',total)
print(All_movie_review.head())

['C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2010.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2011.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2012.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2013.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2014.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2015_2017.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2018.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Final Dataset\\Review_2019.csv', 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundati

# Annotate the Dataset with the Gold Label and Sentiment Score

In [4]:
def map_gold_label(rating):
    if rating == 5:
        return 'Neutral'
    elif rating < 5:
        return 'Negative'
    else:
        return 'Positive'

def map_Sentiment(label):
    if label == 'Positive':
        return 1
    elif label == 'Negative':
        return -1
    else:
        return 0

In [5]:
All_movie_review['Gold_Label']=[ map_gold_label(x) for x in All_movie_review['Rating']]
All_movie_review['Sentiment']=[ map_Sentiment(x) for x in All_movie_review['Gold_Label']]
All_movie_review.head()

Movie_Name                                       Review_Title  \
0  The Social Network                        Bumper sticker philosophies   
1  The Social Network                                      Underwhelming   
2  The Social Network                     My Masturbation Over a Trailer   
3  The Social Network  I hate Facebook but this film is the best of t...   
4  The Social Network  Just a great story of something we now take fo...   

    Review_User                                     Review_content  Rating  \
0       brsmarc  While its never satisfying to write a review a...     NaN   
1       lisamvm  I am a HUGE fan of all of Aaron Sorkins work I...     5.0   
2      paperpag  When I first saw the trailer I couldnt believe...     5.0   
3      chujfugh  No other film I feel captures the basement dwe...    10.0   
4  Filmaholic78  I give this film 10 out of 10Im not sure how m...    10.0   

  Gold_Label  Sentiment  
0   Positive          1  
1    Neutral          0  
2    Neutral          0  
3   Positive          1  
4   Positive          1

# Pre-process the Dataset for Vectorization

In [6]:
# Convert the review content to a Tokenized text.
from gensim.utils import simple_preprocess
All_movie_review['tokenized_text'] = [simple_preprocess(line, deacc=True) for line in All_movie_review['Review_content']]
print(All_movie_review['tokenized_text'].head(10))

0    [while, its, never, satisfying, to, write, rev...
1    [am, huge, fan, of, all, of, aaron, sorkins, w...
2    [when, first, saw, the, trailer, couldnt, beli...
3    [no, other, film, feel, captures, the, basemen...
4    [give, this, film, out, of, im, not, sure, how...
5    [had, heard, plenty, of, positive, comments, a...
6    [it, is, surprising, and, quite, shocking, for...
7    [mark, zuckerberg, is, an, asshat, that, made,...
8    [did, not, care, very, much, for, this, film, ...
9    [being, in, the, computer, biz, was, intereste...
Name: tokenized_text, dtype: object


In [7]:
# Lemmatize the Tokenized review Content
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
All_movie_review["lemmatized_tokens"] = [[lemmatizer.lemmatize(word) for word in tokens] for tokens in All_movie_review['tokenized_text']]
All_movie_review["lemmatized_tokens"].head(10)

0    [while, it, never, satisfying, to, write, revi...
1    [am, huge, fan, of, all, of, aaron, sorkins, w...
2    [when, first, saw, the, trailer, couldnt, beli...
3    [no, other, film, feel, capture, the, basement...
4    [give, this, film, out, of, im, not, sure, how...
5    [had, heard, plenty, of, positive, comment, ab...
6    [it, is, surprising, and, quite, shocking, for...
7    [mark, zuckerberg, is, an, asshat, that, made,...
8    [did, not, care, very, much, for, this, film, ...
9    [being, in, the, computer, biz, wa, interested...
Name: lemmatized_tokens, dtype: object

In [8]:
# Remove the punctuation if any and get the clean Tokens
import nltk
from nltk.corpus import stopwords
import string

def remove_punct(text):
    no_punct = [c for c in text if c not in string.punctuation]
    return no_punct

All_movie_review['Clean_Tokens'] = [remove_punct(w) for w in All_movie_review["lemmatized_tokens"]] 
print(All_movie_review['Clean_Tokens'])

0        [while, it, never, satisfying, to, write, revi...
1        [am, huge, fan, of, all, of, aaron, sorkins, w...
2        [when, first, saw, the, trailer, couldnt, beli...
3        [no, other, film, feel, capture, the, basement...
4        [give, this, film, out, of, im, not, sure, how...
                               ...                        
55736    [marshall, goldsmith, deserves, alot, of, resp...
55737    [since, pioneering, the, degree, assessment, i...
55738    [marshall, goldsmith, is, my, role, model, and...
55739    [my, family, and, watched, the, anniversary, f...
55740    [excellent, movie, about, greek, revolution, l...
Name: Clean_Tokens, Length: 55741, dtype: object


# Split the Dataset for Training and Test

In [9]:
# Split the DataSet into Training and Test with 80-20 composition
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(All_movie_review[['Movie_Name','Review_Title','Review_User','Review_content', 'Gold_Label','lemmatized_tokens','Clean_Tokens']], 
                                                        All_movie_review['Sentiment'],test_size=0.2,random_state=15)
 
x_train =x_train.reset_index()
x_test = x_test.reset_index()
y_train =y_train.to_frame()
y_train = y_train.reset_index()
y_test = y_test.to_frame()
y_test = y_test.reset_index()
print("Number of Records In Training Dataset:", len(x_train))
print("Number of Records In Test Dataset:", len(x_test))


# Build the Movie Review Data Dictionary

In [10]:
from gensim import corpora
movie_review_dict= corpora.Dictionary(All_movie_review["Clean_Tokens"])
print("Total unique words:\n")
print(len(movie_review_dict.token2id))

Total unique words:

165696


# Method 1 -TfIDF Vectorization

In [15]:
# 2 - Vectorize using Tfidf
from gensim.models import TfidfModel
corpus = [movie_review_dict.doc2bow(line) for line in All_movie_review['lemmatized_tokens']]

# TF-IDF Model
tfidf_model = TfidfModel(corpus)

# Vectorize the Training Dataset and Store it in a file

**Time taken to create tfidf for :5178.276551246643**

In [20]:
import gensim
import time
start_time = time.time()
vocab_len = len(movie_review_dict.token2id)
train_features_tfidf = []
tfidf_filename = 'Movie_review_tfidf.csv'
with open(tfidf_filename, 'w+', encoding = 'utf8') as tfidf_file:
    for index, row in x_train.iterrows():
        doc=movie_review_dict.doc2bow(row['lemmatized_tokens'])
        features = gensim.matutils.corpus2csc([tfidf_model[doc]], num_terms=vocab_len).toarray()[:,0]
        train_features_tfidf.append(features)
        if index == 0:
            header = ",".join(str(movie_review_dict[ele]) for ele in range(vocab_len))
            tfidf_file.write(header)
            tfidf_file.write("\n")
        line1 = ",".join( [str(vector_element) for vector_element in features] )
        tfidf_file.write(line1)
        tfidf_file.write('\n')
print("Time taken to create tfidf for :" + str(time.time() - start_time))

Time taken to create tfidf for :5178.276551246643


# Fit the Training TfIDF vectorized dataset for Decision tree Classifier
 **Since the Size of the Vectorization File is More than 27GB I am getting an Memory error while Fit the Training Dataset for Decision Tree Classifier. So As I discussed with the professor I am going to Continue My analysis with Top 25 Reviews.**

In [16]:
from sklearn.tree import DecisionTreeClassifier
import time
start_time = time.time()

# # Read the TFIDF vectors
tfidf_df = pd.read_csv('Movie_review_tfidf.csv')

# Initialize the model
clf_decision_tfidf = DecisionTreeClassifier(random_state=2)

# Fit the model 
clf_decision_tfidf.fit(tfidf_df, y_train['Sentiment'])
print("Time to taken to fit the TF-IDF as input for classifier: " + str(time.time() - start_time))

MemoryError: 

# Import All Movies Top 25 review For NLP 

In [2]:
filepath = 'C:/Users/Vimalathithan/Visha/MS details/Spring 2021/Data Foundation/Final Project/Review_All_Movie_top25.csv'
All_movie_review_top25 = pd.read_csv(filepath,encoding='utf-8',header=0,usecols=['Movie_Name','Review_Title','Review_User','Review_content','Rating'])
print(len(All_movie_review_top25))

16515


In [5]:
# Annotate the Dataset with the Gold Label and Sentiment Score
All_movie_review_top25['Gold_Label']=[ map_gold_label(x) for x in All_movie_review_top25['Rating']]
All_movie_review_top25['Sentiment']=[ map_Sentiment(x) for x in All_movie_review_top25['Gold_Label']]
All_movie_review_top25.head()

Movie_Name                                       Review_Title  \
0  The Wolf of Wall Street                                       Oscar worthy   
1  The Wolf of Wall Street            Awesome movie watched it multiple times   
2  The Wolf of Wall Street  Money is the oxygen of capitalism and I wanna ...   
3  The Wolf of Wall Street  DiCaprio gives the best performance of his career   
4  The Wolf of Wall Street                Stylish funny and very entertaining   

     Review_User                                     Review_content  Rating  \
0  karolin-99263  The movie Leonardo DiCaprio should have gotten...    10.0   
1   MR_Heraclius  This film is every definition of the word impe...    10.0   
2  notoriousCASK  Its no surprise that Scorsese is associated wi...     9.0   
3   BrentHankins  In the mid1990s Jordan Belfort Leonardo DiCapr...     9.0   
4      aaron1506  The Wolf of Wall Street is Scorseses best film...     8.0   

  Gold_Label  Sentiment  
0   Positive          1  
1   Positive          1  
2   Positive          1  
3   Positive          1  
4   Positive          1

# Pre-Process the Dataset

In [6]:
# Convert the review content to a Tokenized text.
from gensim.utils import simple_preprocess
# Tokenize the text column to get the new column 'tokenized_text'
All_movie_review_top25['tokenized_text'] = [simple_preprocess(line, deacc=True) for line in All_movie_review_top25['Review_content']]
print(All_movie_review_top25['tokenized_text'].head(10))

0    [the, movie, leonardo, dicaprio, should, have,...
1    [this, film, is, every, definition, of, the, w...
2    [its, no, surprise, that, scorsese, is, associ...
3    [in, the, mid, jordan, belfort, leonardo, dica...
4    [the, wolf, of, wall, street, is, scorseses, b...
5    [this, film, arrived, in, the, uk, with, award...
6    [brilliantly, acted, superbly, written, and, a...
7    [you, can, literally, watch, any, movie, for, ...
8    [martin, scorsese, has, done, it, again, his, ...
9    [perfectly, executed, in, almost, all, aspects...
Name: tokenized_text, dtype: object


In [7]:
# Lemmatize the Tokenized review Content
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
All_movie_review_top25["lemmatized_tokens"] = [[lemmatizer.lemmatize(word) for word in tokens] for tokens in All_movie_review_top25['tokenized_text']]
All_movie_review_top25["lemmatized_tokens"].head(10)

0    [the, movie, leonardo, dicaprio, should, have,...
1    [this, film, is, every, definition, of, the, w...
2    [it, no, surprise, that, scorsese, is, associa...
3    [in, the, mid, jordan, belfort, leonardo, dica...
4    [the, wolf, of, wall, street, is, scorsese, be...
5    [this, film, arrived, in, the, uk, with, award...
6    [brilliantly, acted, superbly, written, and, a...
7    [you, can, literally, watch, any, movie, for, ...
8    [martin, scorsese, ha, done, it, again, his, n...
9    [perfectly, executed, in, almost, all, aspect,...
Name: lemmatized_tokens, dtype: object

# Split the Dataset into Training and Test

In [8]:
# Split the DataSet into Training and Test with 80-20 composition
from sklearn.model_selection import train_test_split

x_train_top25,x_test_top25,y_train_top25,y_test_top25 = train_test_split(All_movie_review_top25[['Movie_Name','Review_Title','Review_User','Review_content', 'Gold_Label','lemmatized_tokens']], 
                                                        All_movie_review_top25['Sentiment'],test_size=0.2,random_state=15)
 
x_train_top25 =x_train_top25.reset_index()
x_test_top25 = x_test_top25.reset_index()
y_train_top25 =y_train_top25.to_frame()
y_train_top25 = y_train_top25.reset_index()
y_test_top25 = y_test_top25.to_frame()
y_test_top25 = y_test_top25.reset_index()
print("Number of Records In Training Dataset:", len(x_train_top25))
print("Number of Records In Test Dataset:", len(x_test_top25))

Number of Records In Training Dataset: 13212
Number of Records In Test Dataset: 3303


# Build the word Dictionary

In [9]:
from gensim import corpora
movie_review_top25_dict= corpora.Dictionary(All_movie_review_top25["lemmatized_tokens"])
print("Total unique words:\n")
print(len(movie_review_top25_dict.token2id))

Total unique words:

87666


# Method 1 -TfIDF Vectorization

In [12]:
# 1 - TF-IDF Model
from gensim.models import TfidfModel
corpus_top25 = [movie_review_top25_dict.doc2bow(line) for line in All_movie_review_top25['lemmatized_tokens']]
tfidf_model_top25 = TfidfModel(corpus_top25)

# Vectorize the Training Dataset and Store it in a file
Time taken to create tfidf for :709.7966206073761

In [27]:
import gensim
import time
start_time = time.time()
vocab_len_top25 = len(movie_review_top25_dict.token2id)
train_features_tfidf_top25 = []
tfidf_filename = 'Movie_review_top25_tfidf.csv'
with open(tfidf_filename, 'w+', encoding = 'utf8') as tfidf_file:
    for index, row in x_train_top25.iterrows():
        doc=movie_review_top25_dict.doc2bow(row['lemmatized_tokens'])
        features = gensim.matutils.corpus2csc([tfidf_model_top25[doc]], num_terms=vocab_len_top25).toarray()[:,0]
        train_features_tfidf_top25.append(features)
        if index == 0:
            header = ",".join(str(movie_review_top25_dict[ele]) for ele in range(vocab_len_top25))
            tfidf_file.write(header)
            tfidf_file.write("\n")
        line1 = ",".join( [str(vector_element) for vector_element in features] )
        tfidf_file.write(line1)
        tfidf_file.write('\n')
print("Time taken to create tfidf for :" + str(time.time() - start_time))

Time taken to create tfidf for :709.7966206073761


# Training Sentiment Classification Model using TFIDF Vectors
Time to taken to fit the TF-IDF as input for classifier: 9110.061611652374


In [13]:
from sklearn.tree import DecisionTreeClassifier
import time
start_time = time.time()

# Read the TFIDF vectors
tfidf_top25_df = pd.read_csv('Movie_review_top25_tfidf.csv')

# Initialize the model
clf_decision_top25_tfidf = DecisionTreeClassifier(random_state=2)

# Fit the model 
clf_decision_top25_tfidf.fit(tfidf_top25_df, y_train_top25['Sentiment'])
print("Time to taken to fit the TF-IDF as input for classifier: " + str(time.time() - start_time))


Time to taken to fit the TF-IDF as input for classifier: 9110.061611652374


# Test the TF-IDF Model for Test Dataset
Time taken to predict using TF-IDF:8.70795488357544

In [17]:
import gensim
import time
from sklearn.metrics import classification_report
vocab_len_top25 = len(movie_review_top25_dict.token2id)
test_features_top25_tfidf = []
start_time = time.time()
for index, row in x_test_top25.iterrows():
    doc = movie_review_top25_dict.doc2bow(row['lemmatized_tokens'])
    features = gensim.matutils.corpus2csc([tfidf_model_top25[doc]], num_terms=vocab_len_top25).toarray()[:,0]
    test_features_top25_tfidf.append(features)

# Use the model's predict method
test_predictions_top25_tfidf = clf_decision_top25_tfidf.predict(test_features_top25_tfidf)
print(classification_report(y_test_top25['Sentiment'],test_predictions_top25_tfidf))
print("Time taken to predict using TF-IDF:" + str(time.time() - start_time))

              precision    recall  f1-score   support

          -1       0.40      0.40      0.40       669
           0       0.08      0.05      0.06       204
           1       0.81      0.83      0.82      2430

    accuracy                           0.69      3303
   macro avg       0.43      0.43      0.43      3303
weighted avg       0.68      0.69      0.68      3303

Time taken to predict using TF-IDF:8.70795488357544


# TFIDF ACCURACY = 69%
Classification Report shows the average accuracy which is **0.69** . This is a good result compared to the amount of data used for training. predict function can be used on the model object to get the predicted class for the test data. Accuracy for positive and negative sentiments is better than neutral which makes sense as it is hard to distinguish the neutral comments compared to commonly used words in the positive and negative sentiment.

# Method 2 - Word2vec Skip gram vectorization

In [18]:
# 2 - Vectorize using word2vec
from gensim import corpora
from gensim.models import Word2Vec
features=[w for w in All_movie_review_top25['lemmatized_tokens']]
start_time = time.time()
word2vecmodel = Word2Vec(features,size=1000,window=3,min_count=1,workers=3,sg=1)
print (word2vecmodel)

Word2Vec(vocab=87666, size=1000, alpha=0.025)


# Vectorize the Training Dataset and Store it in a file

In [27]:
import gensim
import time
start_time = time.time()
word2vec_top25_filename = 'Movie_review_top25_word2vec.csv'
with open(word2vec_top25_filename, 'w+', encoding = 'utf8') as word2vec_file:
    for index, row in x_train_top25.iterrows():
#         word2vec_vector = (np.mean([word2vecmodel[token] for token in row['lemmatized_tokens'], axis=0)).tolist()
        word2vec_vector = (np.mean([word2vecmodel[token] for token in row['lemmatized_tokens']], axis=0)).tolist()
        if index == 0:
            header = ",".join(str(ele) for ele in range(1000))
            word2vec_file.write(header)
            word2vec_file.write("\n")
        line1 = ",".join( [str(vector_element) for vector_element in word2vec_vector] )
        word2vec_file.write(line1)
        word2vec_file.write('\n')
print("Time taken to create word2vec for :" + str(time.time() - start_time))

<ipython-input-27-01221f59ac7c>:8: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  word2vec_vector = (np.mean([word2vecmodel[token] for token in row['lemmatized_tokens']], axis=0)).tolist()


Time taken to create word2vec for :42.52248930931091


# Training Sentiment Classification Model using Word2Vec Vectors

In [28]:
import time
#Import the DecisionTreeeClassifier
from sklearn.tree import DecisionTreeClassifier
# Load from the filename
word2vec_top25_df = pd.read_csv(word2vec_top25_filename)
#Initialize the model
clf_decision_word2vec_top25 = DecisionTreeClassifier()

start_time = time.time()
# Fit the model
clf_decision_word2vec_top25.fit(word2vec_top25_df, y_train_top25['Sentiment'])
print("Time taken to fit the model with word2vec vectors: " + str(time.time() - start_time))

Time taken to fit the model with word2vec vectors: 25.634369134902954


# Test the Model for Word2Vec Vectorization

In [31]:
from sklearn.metrics import classification_report
test_features_top25_word2vec = []
import time
start_time = time.time()
for index, row in x_test_top25.iterrows():
    word2vec_vector = (np.mean([word2vecmodel[token] for token in row['lemmatized_tokens']], axis=0)).tolist()
    if type(word2vec_vector) is list:
        test_features_top25_word2vec.append(word2vec_vector)
    else:
        test_features_top25_word2vec.append(np.array([0 for i in range(1000)]))
# Use the model's predict method
test_predictions_top25_word2vec = clf_decision_word2vec_top25.predict(test_features_top25_word2vec)
print(classification_report(y_test_top25['Sentiment'],test_predictions_top25_word2vec))
print("Time taken to predict using Word2Vec:" + str(time.time() - start_time))

<ipython-input-31-3b90746239f2>:6: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  word2vec_vector = (np.mean([word2vecmodel[token] for token in row['lemmatized_tokens']], axis=0)).tolist()


              precision    recall  f1-score   support

          -1       0.38      0.40      0.39       669
           0       0.10      0.11      0.10       204
           1       0.81      0.80      0.80      2430

    accuracy                           0.67      3303
   macro avg       0.43      0.43      0.43      3303
weighted avg       0.68      0.67      0.68      3303

Time taken to predict using Word2Vec:5.687241315841675


# Word2Vec ACCURACY = 67%
Classification Report shows the average accuracy which is **0.67** . This is a good result compared to the amount of data used for training. predict function can be used on the model object to get the predicted class for the test data. Accuracy for positive and negative sentiments is better than neutral which makes sense as it is hard to distinguish the neutral comments compared to commonly used words in the positive and negative sentiment.

Compared to TFIDF vectorisation the accuracy is reduced by 2% eventhough it is a good result compared to number of datas used in the training.In the performance perspective , Word2Vec is better than the TF-IDF vectorisation.

# Method 3- BOW Vectorization

In [32]:
import time
start_time = time.time()
# vocab_len_top25 = len(movie_review_top25_dict.token2id)
bowvec_top25_filename = 'Movie_review_top25_bowvec.csv'
with open(bowvec_top25_filename, 'w+', encoding='utf8') as bow_file:
    for index, row in x_train_top25.iterrows():
        features = gensim.matutils.corpus2csc([movie_review_top25_dict.doc2bow(row['lemmatized_tokens'])],num_terms=vocab_len_top25).toarray()[:,0]
        if index == 0:
            #print("Header")
            header = ",".join(str(movie_review_top25_dict[ele]) for ele in range(vocab_len_top25))
            #print(header)
            bow_file.write(header)
            bow_file.write("\n")
        line1 = ",".join( [str(vector_element) for vector_element in features] )
        bow_file.write(line1)
        bow_file.write('\n')

print("Time taken to create bow for :" + str(time.time() - start_time))

Time taken to create bow for :647.8690655231476


In [11]:
from sklearn.tree import DecisionTreeClassifier
import time
bowvec_top25_df = pd.read_csv('Movie_review_top25_bowvec.csv')
# Train the classifier with default parameters
start_time = time.time()
# Initialize the classifier object
bow_top25_clf = DecisionTreeClassifier(random_state=0)
# Fit the model with input vectors and corresponding sentiment labels
bow_top25_clf.fit(bowvec_top25_df, y_train_top25['Sentiment'])
print("Time taken to fit the model: " + str(time.time() - start_time))

Time taken to fit the model: 195.21626496315002


# Test the BOW Model

In [14]:
import gensim
from sklearn.metrics import classification_report
# Iterating through test data to get the predictions of sentiment by the model
vocab_len_top25 = len(movie_review_top25_dict.token2id)
test_bow_top25_features = []
start_time = time.time()
for index, row in x_test_top25.iterrows():
    # Converting the tokens into the format that the model requires
    features = gensim.matutils.corpus2csc([movie_review_top25_dict.doc2bow(row['lemmatized_tokens'])],num_terms=vocab_len_top25).toarray()[:,0]
    test_bow_top25_features.append(features)
test_bow_top25_predictions = bow_top25_clf.predict(test_bow_top25_features)
# Comparing the predictions to actual sentiments for the sentences
print(classification_report(y_test_top25['Sentiment'],test_bow_top25_predictions))
print("Time taken to predict the model " + str(time.time() - start_time))

              precision    recall  f1-score   support

          -1       0.46      0.46      0.46       669
           0       0.12      0.07      0.09       204
           1       0.81      0.84      0.83      2430

    accuracy                           0.72      3303
   macro avg       0.47      0.46      0.46      3303
weighted avg       0.70      0.72      0.71      3303

Time taken to predict the model 7.7208263874053955


# BOW Vectorization ACCURACY = 72%
Classification Report shows the average accuracy which is **0.72** . This is a very good result compared to the amount of data used for training and also to the previous vectorisation. predict function can be used on the model object to get the predicted class for the test data. Accuracy for positive and negative sentiments is better than neutral which makes sense as it is hard to distinguish the neutral comments compared to commonly used words in the positive and negative sentiment.

Compared to TFIDF, word2vec vectorisation the accuracy is increased. And also by comparing the performance BOW vectorisation took less time for fit and test the model.

# Method 4 - Document Embeddings (Doc2Vec)

In [15]:
from gensim.models.doc2vec import Doc2Vec, TaggedDocument

# TaggedDocuments are tuple of stemmed_tokens and class lable, example is printed (scroll to the right to see label)
documents = [TaggedDocument(doc, [i]) for i, doc in enumerate(All_movie_review_top25['lemmatized_tokens'])]
print(documents[1])

# Initialize the model (try also with min_count = 2,3,etc...)
doc2vec_model = Doc2Vec(documents, vector_size=1000, window=3, min_count=1, workers=4)

# Sample vector for the stemmed tokens
vector = doc2vec_model.infer_vector(All_movie_review_top25['lemmatized_tokens'][0])
# Printing sample vector
print(len(vector))
print("Top 10 values in Doc2Vec inferred vector:")
print(vector[:10])

TaggedDocument(['this', 'film', 'is', 'every', 'definition', 'of', 'the', 'word', 'impeccable', 'scorceses', 'fantastic', 'storytelling', 'ability', 'mixed', 'with', 'phenomenal', 'acting', 'from', 'dicaprio', 'hill', 'and', 'the', 'rest', 'of', 'the', 'cast', 'make', 'for', 'an', 'awesome', 'combination', 'the', 'film', 'is', 'never', 'boring', 'and', 'give', 'you', 'glimpse', 'into', 'the', 'life', 'of', 'man', 'who', 'had', 'everything', 'in', 'the', 'world', 'only', 'to', 'have', 'it', 'taken', 'away'], [1])
1000
Top 10 values in Doc2Vec inferred vector:
[ 0.00669052 -0.00554525  0.02067334  0.01022999 -0.00792841  0.00429593
  0.03000394 -0.00170983  0.02583175  0.00547256]


In [16]:
import time
start_time = time.time()
# Store the vectors for train data in 
doc2vec_top25_filename = 'Movie_review_top25_doc2vec.csv'
with open(doc2vec_top25_filename, 'w+') as doc2vec_file:
    for index, row in x_train_top25.iterrows():
        model_vector = doc2vec_model.infer_vector(row['lemmatized_tokens'])
        if index == 0:
            header = ",".join(str(ele) for ele in range(1000))
            doc2vec_file.write(header)
            doc2vec_file.write("\n")
        line1 = ",".join( [str(vector_element) for vector_element in model_vector] )
        doc2vec_file.write(line1)
        doc2vec_file.write('\n')
print("Time taken to create doc2vec vectors: " + str(time.time() - start_time))

Time taken to create doc2vec vectors: 89.43428301811218


In [19]:
# Load from the filename
from sklearn.tree import DecisionTreeClassifier
import time
start_time = time.time()
doc2vec_top25_df = pd.read_csv(doc2vec_top25_filename)

clf_decision_top25_doc2vec = DecisionTreeClassifier()
# Fit the model
clf_decision_top25_doc2vec.fit(doc2vec_top25_df, y_train_top25['Sentiment'])
print("Time taken to fit the model: " + str(time.time() - start_time))

Time taken to fit the model: 25.365626335144043


In [21]:
from sklearn.metrics import classification_report
import time
start_time = time.time()
test_features_top25_doc2vec = []
for index, row in x_test_top25.iterrows():
    model_vector = doc2vec_model.infer_vector(row['lemmatized_tokens'])
    test_features_top25_doc2vec.append(model_vector)
test_predictions_top25_doc2vec = clf_decision_top25_doc2vec.predict(test_features_top25_doc2vec)
print(classification_report(y_test_top25['Sentiment'],test_predictions_top25_doc2vec))
print("Time taken to predict the model " + str(time.time() - start_time))

              precision    recall  f1-score   support

          -1       0.32      0.33      0.33       669
           0       0.08      0.09      0.09       204
           1       0.78      0.77      0.77      2430

    accuracy                           0.64      3303
   macro avg       0.39      0.40      0.40      3303
weighted avg       0.65      0.64      0.64      3303

Time taken to predict the model 16.563765048980713


# Doc2vec ACCURACY = 64%
Classification Report shows the average accuracy which is **0.64** . This is a very good result compared to the amount of data used for training and also to the previous vectorisation. predict function can be used on the model object to get the predicted class for the test data. Accuracy for positive and negative sentiments is better than neutral which makes sense as it is hard to distinguish the neutral comments compared to commonly used words in the positive and negative sentiment.

Compared to TFIDF, word2vec,BOW vectorisation the accuracy is increased. And also by comparing the performance Doc2vec vectorisation took less time for fit and test the model.

# Conclusion:
*Based on my Text analysis related to Biographic movie I come to concusion that BOW Vectorization and Doc2Vec Document Embedding will provide a great accuracy in predicting the test Data . At the same Vectorisation took less time for training and test the model.*